File: wpilibcExamples/src/main/cpp/examples/StateSpaceSingleJointedArm/cpp/subsystems/SingleJointedArmCoeffs.cpp
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2019-2020 FIRST. All Rights Reserved.                        */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

#include "subsystems/SingleJointedArmCoeffs.h"

#include <Eigen/Core>

frc::StateSpacePlantCoeffs<2, 1, 1> MakeSingleJointedArmPlantCoeffs() {
  Eigen::Matrix<double, 2, 2> A;
  A(0, 0) = 1.0;
  A(0, 1) = 0.005049987902346862;
  A(1, 0) = 0.0;
  A(1, 1) = 0.9999952088540683;
  Eigen::Matrix<double, 2, 1> B;
  B(0, 0) = 1.1444063060619785e-06;
  B(1, 0) = 0.0004532298583066335;
  Eigen::Matrix<double, 1, 2> C;
  C(0, 0) = 1.0;
  C(0, 1) = 0.0;
  Eigen::Matrix<double, 1, 1> D;
  D(0, 0) = 0.0;
  return frc::StateSpacePlantCoeffs<2, 1, 1>(A, B, C, D);
}

frc::StateSpaceControllerCoeffs<2, 1, 1>
MakeSingleJointedArmControllerCoeffs() {
  Eigen::Matrix<double, 1, 2> K;
  K(0, 0) = 659.4457876860411;
  K(0, 1) = 179.11558938712352;
  Eigen::Matrix<double, 1, 2> Kff;
  Kff(0, 0) = 0.5390980036447967;
  Kff(0, 1) = 8.538202281633405;
  Eigen::Matrix<double, 1, 1> Umin;
  Umin(0, 0) = -12.0;
  Eigen::Matrix<double, 1, 1> Umax;
  Umax(0, 0) = 12.0;
  return frc::StateSpaceControllerCoeffs<2, 1, 1>(K, Kff, Umin, Umax);
}

frc::StateSpaceObserverCoeffs<2, 1, 1> MakeSingleJointedArmObserverCoeffs() {
  Eigen::Matrix<double, 2, 1> K;
  K(0, 0) = 0.8033915976431203;
  K(1, 0) = 7.738121910642457;
  return frc::StateSpaceObserverCoeffs<2, 1, 1>(K);
}

frc::StateSpaceLoop<2, 1, 1> MakeSingleJointedArmLoop() {
  return frc::StateSpaceLoop<2, 1, 1>(MakeSingleJointedArmPlantCoeffs(),
                                      MakeSingleJointedArmControllerCoeffs(),
                                      MakeSingleJointedArmObserverCoeffs());
}

File: wpilibjExamples/src/main/java/edu/wpi/first/wpilibj/examples/statespacesinglejointedarm/subsystems/SingleJointedArmCoeffs.java
package edu.wpi.first.wpilibj.examples.statespacesinglejointedarm.subsystems;

//CHECKSTYLE.OFF: AvoidStarImport
import edu.wpi.first.wpilibj.controller.StateSpaceControllerCoeffs;
import edu.wpi.first.wpilibj.controller.StateSpaceLoop;
import edu.wpi.first.wpilibj.controller.StateSpaceObserverCoeffs;
import edu.wpi.first.wpilibj.controller.StateSpacePlantCoeffs;
import edu.wpi.first.wpiutil.math.*;
import edu.wpi.first.wpiutil.math.numbers.*;
//CHECKSTYLE.ON

public final class SingleJointedArmCoeffs {

  private SingleJointedArmCoeffs() {
  }

  @SuppressWarnings({"LocalVariableName", "JavadocMethod"})
  public static StateSpacePlantCoeffs<N2, N1, N1>
      makeSingleJointedArmPlantCoeffs() {
    Matrix<N2, N2> A = MatrixUtils.mat(Nat.N2(), Nat.N2())
        .fill(1.0, 0.005049987902346862, 0.0, 0.9999952088540683);
    Matrix<N2, N1> B = MatrixUtils.mat(Nat.N2(), Nat.N1())
        .fill(1.1444063060619785e-06, 0.0004532298583066335);
    Matrix<N1, N2> C = MatrixUtils.mat(Nat.N1(), Nat.N2()).fill(1.0, 0.0);
    Matrix<N1, N1> D = MatrixUtils.mat(Nat.N1(), Nat.N1()).fill(0.0);
    return new StateSpacePlantCoeffs<N2, N1, N1>(Nat.N2(), Nat.N1(), Nat.N1(), A, B, C, D);
  }

  @SuppressWarnings({"LocalVariableName", "JavadocMethod"})
  public static StateSpaceControllerCoeffs<N2, N1, N1>
      makeSingleJointedArmControllerCoeffs() {
    Matrix<N1, N2> K = MatrixUtils.mat(Nat.N1(), Nat.N2())
        .fill(659.4457876860411, 179.11558938712352);
    Matrix<N1, N2> Kff = MatrixUtils.mat(Nat.N1(), Nat.N2())
        .fill(0.5390980036447967, 8.538202281633405);
    Matrix<N1, N1> Umin = MatrixUtils.mat(Nat.N1(), Nat.N1()).fill(-12.0);
    Matrix<N1, N1> Umax = MatrixUtils.mat(Nat.N1(), Nat.N1()).fill(12.0);
    return new StateSpaceControllerCoeffs<N2, N1, N1>(K, Kff, Umin, Umax);
  }

  @SuppressWarnings({"LocalVariableName", "JavadocMethod"})
  public static StateSpaceObserverCoeffs<N2, N1, N1>
      makeSingleJointedArmObserverCoeffs() {
    Matrix<N2, N1> K = MatrixUtils.mat(Nat.N2(), Nat.N1())
        .fill(0.8033915976431203, 7.738121910642457);
    return new StateSpaceObserverCoeffs<N2, N1, N1>(K);
  }

  @SuppressWarnings("JavadocMethod")
  public static StateSpaceLoop<N2, N1, N1> makeSingleJointedArmLoop() {
    return new StateSpaceLoop<N2, N1, N1>(makeSingleJointedArmPlantCoeffs(),
                                          makeSingleJointedArmControllerCoeffs(),
                                          makeSingleJointedArmObserverCoeffs());
  }
}
